pkg/helm/watcher/cache/cache.go

```diff
@@ -0,0 +1,182 @@
+package cache
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/ghodss/yaml"
+	"github.com/gofrs/flock"
+
+	pb "github.com/weaveworks/weave-gitops/pkg/api/profiles"
+)
+
+const (
+	// we globally lock the whole base cache folder to avoid the
+	// possibility of checking if a sub-folder exists, just so it is created the moment
+	// we try to create it after we checked that it exists.
+	lockFilename    = "cache.lock"
+	profileFilename = "profiles.yaml"
+	valuesFilename  = "values.yaml"
+)
+
+// type alias for easier reading of the cache layout for values.
+type profileName = string
+type profileVersion = string
+
+// ValueMap contains easy access for a profile name and version based values file.
+type ValueMap map[profileName]map[profileVersion][]byte
+
+// Cache defines an interface to work with the profile data cacher.
+//go:generate go run github.com/maxbrunsfeld/counterfeiter/v6 -generate
+//counterfeiter:generate . Cache
+type Cache interface {
+	Update(helmRepoNamespace, helmRepoName string, value Data) error
+	// GetProfiles specifically retrieve profiles data only to avoid traversing the values structure for no reason.
+	GetProfiles(helmRepoNamespace, helmRepoName string) ([]*pb.Profile, error)
+	// GetProfileValues will try and find a specific values file for the given profileName and profileVersion. Returns an
+	// error if said version is not found.
+	GetProfileValues(helmRepoNamespace, helmRepoName, profileName, profileVersion string) ([]byte, error)
+}
+
+// Data is explicit data for a specific profile including values.
+// Saved as `profiles.yaml` and `profileName/version/values.yaml`.
+type Data struct {
+	Profiles []*pb.Profile `yaml:"profiles"`
+	Values   ValueMap
+}
+
+// HelmCache is used to cache profiles data from scanner helm repositories.
+type HelmCache struct {
+	cacheDir string
+}
+
+var _ Cache = &HelmCache{}
+
+// NewCache initialises the cache and returns it.
+func NewCache(cacheDir string) (*HelmCache, error) {
+	if _, err := os.Stat(cacheDir); os.IsNotExist(err) {
+		if err := os.MkdirAll(cacheDir, 0700); err != nil {
+			return nil, fmt.Errorf("failed to create helm cache dir: %w", err)
+		}
+	}
```

Comment:
```suggestion
	if err := os.MkdirAll(cacheDir, 0700); err != nil {
		return nil, fmt.Errorf("failed to create helm cache dir: %w", err)
	}
```

```
// If path is already a directory, MkdirAll does nothing
// and returns nil.
func MkdirAll(path string, perm FileMode) error {
	// Fast path: if we can tell whether path is a directory or file, stop with success or error.
	dir, err := Stat(path)
	if err == nil {
		if dir.IsDir() {
			return nil
		}
		return &PathError{Op: "mkdir", Path: path, Err: syscall.ENOTDIR}
	}
```

Reply:
Ah thanks for reminding me. :D I had some other logic in there, but removed that.